read_header.cpp

```diff
@@ -87,12 +88,180 @@ int main() {
+    std::vector<std::vector<std::vector<double>>> electron_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));
+    std::vector<std::vector<std::vector<double>>> hydrogen_number_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));
+    std::vector<std::vector<std::vector<double>>> helium_number_density(hinfo.global_nx, std::vector<std::vector<double>>(hinfo.global_ny, std::vector<double>(hinfo.global_nz, 0.0)));
+    // Assuming a constant ratio of He to H for simplicity
+    double He_to_H_ratio = 0.1; // Assuming a constant ratio for simplicity
+    std::cout << "Calculating temperature and electron density and densities of Hydrogen and Helium\n";
+    for (int i = 0; i < hinfo.global_nx; ++i) {
+        for (int j = 0; j < hinfo.global_ny; ++j) {
+            for (int k = 0; k < hinfo.global_nz; ++k) {
+                double rho = phys_var[0][i][j][k]; // Density
+                double e_int = phys_var[4][i][j][k]; // Internal energy density
+                if (rho > 0 && e_int > 0) { // Avoid division by zero
+                    auto temp_pair = boost::math::tools::bisect(
+                        [&](double T) { return residual(rho, T, e_int); },
+                        T_min, T_max,
+                        boost::math::tools::eps_tolerance<double>(30)
+                    );
+                    Temperature[i][j][k] = (temp_pair.first + temp_pair.second) / 2.0;
+                    double nH = rho * X / m_p; // Number density of hydrogen
+                    double mu_val = mu(nH, Temperature[i][j][k]);
+                                    // compute electron density
+	                // N.B. it is absolutely critical to include the metal contribution here!
+	                // the approximation for the metals contribution to e- fails at high densities (~1e3 or higher)
+                    double n_e = (rho / (m_p + m_e)) * (1.0 - mu_val * (X + Y / 4. + Z / mean_metals_A)) / (mu_val - (electron_mass_cgs / (m_p + m_e)));
+                    if (n_e > 1.0e-4 * nH) {
+                        electron_density[i][j][k] = 1.0e-4 * nH; // Set a minimum electron density
+                    }else if (n_e >=0.0) {
+                        electron_density[i][j][k] = n_e; // Set the calculated electron density
+                    } else {
+                        std::cerr << "Negative electron density calculated at (" << i << ", " << j << ", " << k << "). Setting to zero.\n";
+                        electron_density[i][j][k] = 0.0; // Set to zero or handle as needed
+                    }
+                    
+                } else {
+                    std::cerr << "Invalid density or internal energy at (" << i << ", " << j << ", " << k << "). Setting temperature to zero.\n";
```

Comment:
Negative densities and internal energies are handled in the code. We need not check them.